snow/engine/snowman/transitive.go

```diff
@@ -170,23 +174,39 @@ func (t *Transitive) Put(ctx context.Context, nodeID ids.NodeID, requestID uint3
 		return t.GetFailed(ctx, nodeID, requestID)
 	}
 
-	actualBlkID := blk.ID()
-	expectedBlkID, ok := t.blkReqs.GetValue(common.Request{
-		NodeID:    nodeID,
-		RequestID: requestID,
-	})
-	// If the provided block is not the requested block, we need to explicitly
-	// mark the request as failed to avoid having a dangling dependency.
-	if ok && actualBlkID != expectedBlkID {
-		t.Ctx.Log.Debug("incorrect block returned in Put",
-			zap.Stringer("nodeID", nodeID),
-			zap.Uint32("requestID", requestID),
-			zap.Stringer("blkID", actualBlkID),
-			zap.Stringer("expectedBlkID", expectedBlkID),
-		)
-		// We assume that [blk] is useless because it doesn't match what we
-		// expected.
-		return t.GetFailed(ctx, nodeID, requestID)
+	var (
+		req = common.Request{
+			NodeID:    nodeID,
+			RequestID: requestID,
+		}
+		sourceMetric prometheus.Counter
+	)
+	switch expectedBlkID, ok := t.blkReqs.GetValue(req); {
+	case ok:
+		actualBlkID := blk.ID()
+		if actualBlkID != expectedBlkID {
+			t.Ctx.Log.Debug("incorrect block returned in Put",
+				zap.Stringer("nodeID", nodeID),
+				zap.Uint32("requestID", requestID),
+				zap.Stringer("blkID", actualBlkID),
+				zap.Stringer("expectedBlkID", expectedBlkID),
+			)
+			// We assume that [blk] is useless because it doesn't match what we
+			// expected.
+			return t.GetFailed(ctx, nodeID, requestID)
+		}
+
+		sourceMetric = t.blkReqSourceMetric[req]
+	case requestID == constants.GossipMsgRequestID:
+		sourceMetric = t.metrics.providerSource.WithLabelValues(putGossipSource)
+	default:
+		// This can happen if this block was provided to this engine while a Get
+		// request was outstanding. For example, the block may have been locally
+		// built or the node may have received a PushQuery with this block.
+		//
+		// Note: It is still possible this block will be issued here, because
+		// the block may have previously failed verification.
+		sourceMetric = t.metrics.providerSource.WithLabelValues(unknownSource)
```

Comment:
When we observe this - release, we shouldn't see many of these... If we do, that's unexpected.